locksmith/src/websub/jobs/locks.ts

```diff
@@ -1,21 +1,80 @@
+import { Op } from 'sequelize'
+import { networks } from '@unlock-protocol/networks'
 import { Lock } from '../../graphql/datasource'
-import { Hook } from '../../models'
+import { Hook, ProcessedHookItem } from '../../models'
 import { TOPIC_LOCKS } from '../topics'
+import { notifyHook, filterHooksByTopic } from '../helpers'
 
-import { networkMapToFnResult, notifyHook } from '../helpers'
+const FETCH_LIMIT = 25
 
-export async function notifyOfLocks(hooks: Hook[]) {
-  const subscribed = hooks.filter((hook) => {
-    const path = new URL(hook.topic).pathname
-    return TOPIC_LOCKS.test(path)
+async function fetchUnprocessedLocks(network: number, page = 0) {
+  const lockSource = new Lock(network)
+  const locks = await lockSource.getLocks({
+    first: FETCH_LIMIT,
+    skip: page ? page * FETCH_LIMIT : 0,
   })
 
-  const networkToLocksMap = await networkMapToFnResult((network) => {
-    const lockSource = new Lock(network)
-    return lockSource.getLocks({ first: 25 })
+  const lockIds = locks.map((lock: any) => lock.id)
+  const processedLocks = await ProcessedHookItem.findAll({
+    where: {
+      objectId: {
+        [Op.in]: lockIds,
+      },
+    },
   })
-  for (const hook of subscribed) {
-    const data = networkToLocksMap.get(Number(hook.network))
-    notifyHook(hook, data)
+
+  const unprocessedLocks = locks.filter(
+    (lock: any) => !processedLocks.find((item) => item.objectId === lock.id)
+  )
+  return unprocessedLocks
+}
+
+async function notifyHooksOfAllUnprocessedLocks(
+  hooks: Hook[],
+  network: number
+) {
+  let page = 0
+  while (true) {
+    const locks = await fetchUnprocessedLocks(network, page)
+
+    // If empty, break the loop and return as there are no more new locks to process.
+    if (!locks.length) {
+      break
+    }
+
+    await Promise.all(
+      hooks.map(async (hook) => {
+        const data = locks
+        const hookEvent = await notifyHook(hook, {
+          data,
+          network,
+        })
+        return hookEvent
+      })
+    )
+
+    const processedHookItems = locks.map((lock: any) => {
+      return {
+        network,
+        type: 'lock',
+        objectId: lock.id,
+      }
+    })
+
+    await ProcessedHookItem.bulkCreate(processedHookItems)
+
+    page += 1
+  }
+}
+
+export async function notifyOfLocks(hooks: Hook[]) {
+  const subscribedHooks = filterHooksByTopic(hooks, TOPIC_LOCKS)
```

Comment:
Same as on the other PR let's add a note to retrieve the hooks *inside* the `notifyHooksOfAllUnprocessedLocks` function so that we only retrieve hooks if/when there applicable ones.